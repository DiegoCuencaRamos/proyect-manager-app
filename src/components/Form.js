import React, { useState } from 'react'
import moment from 'moment'
// React dates
import { DateRangePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css' // CSS
import 'react-dates/initialize' // JS

const ItemForm = ({ isProyect, proyectId, item, onParentFormSubmit }) => {
    // State
    const [ name, setName ] = useState(item ? item.name : '')
    const [ status, setStatus ] = useState(item ? item.status : '')
    const [ invoice, setInvoice ] = useState(item ? item.invoice : '')
    const [ startDate, setStartDate ] = useState(item ? moment(item.startDate) : moment())
    const [ endDate, setEndDate ] = useState(item ? moment(item.endDate) : moment())
    const [ color, setColor ] = useState(item ? item.color : '')
    const [ description, setDescription ] = useState(item ? item.description : '')
    const [ calendarFocused, setCalendarFocused ] = useState(null)
    const [ errorMessage, setErrorMessage ] = useState('')

    const statusStyle = { 
        flexBasis: isProyect ? '48%' : '100%',
        marginRight: isProyect ? '4%' : '0',
    }

    // Events
    const onNameChange = (e) => {
        setName(e.target.value)
    }

    const onStatusChange = (e) => {
        setStatus(e.target.value)
    }

    const onIvoiceChange = (e) => {
        setInvoice(e.target.value)
    }

    const onDatesChange = ({ startDate, endDate }) => {
        setStartDate(startDate)
        setEndDate(endDate)
    }

    const onFocusChange = calendarFocused => {
        setCalendarFocused(calendarFocused)
    }

    const onColorChange = (e) => {
        setColor(e.target.value)
    }

    const onDescriptionChange = (e) => {
        setDescription(e.target.value)
    }

    const onChildFromSubmit = (e) => {
        e.preventDefault()

        const commonItem = {
            name,
            status,
            startDate: startDate.startOf('day').valueOf(),
            endDate: endDate.startOf('day').valueOf(),
            color,
            description,
        }
        
        if (name && status && startDate && endDate && color) {
            setErrorMessage('')
            
            onParentFormSubmit((
                isProyect 
                ? {
                    ...commonItem,
                    invoice
                } : {
                    ...commonItem,
                    proyectId
                }
            )) 
        } else {
            setErrorMessage('Please provide title, status, dates and color for your proyect')
        }
    }

    // Render
    return (
        <div className="container">
            {errorMessage && <p className="error-message">{errorMessage}</p>}

            <form className="form" onSubmit={onChildFromSubmit}>
                <div className="form__wrapper">
                    <input 
                        className="form__item--name"
                        type="text"
                        value={name}
                        placeholder="Name"
                        onChange={onNameChange}
                    />

                    <select 
                        className="form__item--status"
                        style={statusStyle}
                        value={status}
                        onChange={onStatusChange}
                    >
                        <option defaultValue value="">Select status</option>
                        <option value="todo">To do</option>
                        <option value="doing">Doing</option>
                        <option value="done">Done</option>
                    </select>

                    {isProyect && 
                        <input 
                            className="form__item--invoice"
                            type="number"
                            value={invoice}
                            placeholder="Invoice"
                            onChange={onIvoiceChange}
                        />
                    }
                    
                    <DateRangePicker 
                        startDate={startDate}
                        startDateId={startDate.valueOf().toString()}
                        endDate={endDate}
                        endDateId={endDate.valueOf().toString()}
                        onDatesChange={onDatesChange}
                        focusedInput={calendarFocused}
                        onFocusChange={onFocusChange}
                        showClearDates={true}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />

                    <label className="form__item--color"> 
                        {isProyect ? 'Proyect' : 'Task'} color:
                        <input 
                            type="color" 
                            value={color}
                            onChange={onColorChange}
                        />
                    </label>

                    <textarea 
                        className="form__item--description"
                        value={description}
                        placeholder="Description"
                        onChange={onDescriptionChange}
                    />

                    <input 
                        className="form__item--submit button"
                        type="submit"
                        value={`Save ${isProyect ? 'proyect' : 'task'}`}
                    />
                </div>             
            </form>
        </div>
    )
}

export default ItemForm