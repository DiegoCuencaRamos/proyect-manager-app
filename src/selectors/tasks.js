const getVisibleTasks = (proyectId, tasks, { searchText, sortBy }) => {
    return tasks
        // Get filtered tasks by proyect
        .filter(task => task.proyectId === proyectId)
        // Get filtered tasks by searchText
        .filter(task => task.name.toLowerCase().includes(searchText.toLowerCase()))
        // Order tasks by criteria
        .sort((a, b) => {
            if(sortBy === 'name') {
                return a.name.toLowerCase() > b.name.toLowerCase() ? 1 : -1
            }
        }) 
}

export default getVisibleTasks