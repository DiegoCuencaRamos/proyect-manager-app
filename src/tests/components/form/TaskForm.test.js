import React from 'react'
import { mount, shallow } from 'enzyme'
import moment from 'moment'
import Form from '../../../components/Form'
import { tasks } from '../../fixures/tasks'
import { DateRangePicker } from 'react-dates'
import 'react-dates/initialize' // datePicker JS
import { MemoryRouter } from 'react-router'


let wrapper, item, onParentFormSubmit 
const { name, status, startDate, endDate, color, description } = tasks[0]

beforeEach(() => {
    item = { name, status, startDate, endDate, color, description }
    onParentFormSubmit = jest.fn()
})

test('should render Task Form with empty values', () => {
    wrapper = shallow(
        <Form 
            onParentFormSubmit={onParentFormSubmit}
        />
    )
    expect(wrapper).toMatchSnapshot()
})

describe('should render Task Form with task data:', () => {
    beforeEach(() => {
        wrapper = shallow(
            <Form 
                item={item} 
                onParentFormSubmit={onParentFormSubmit}
            />
        )
    })
    
    test(('should set task name input value'), () => {
        expect(wrapper.find('.form__item--name').prop('value')).toBe(item.name)
    })

    test(('should set task status input value'), () => {
        expect(wrapper.find('.form__item--status').prop('value')).toBe(item.status)
    })

    test(('should set task start date input value'), () => {
        expect(wrapper.find(DateRangePicker).prop('startDate')).toEqual(moment(item.startDate))
    })

    test(('should set task end date input value'), () => {
        expect(wrapper.find(DateRangePicker).prop('endDate')).toEqual(moment(item.endDate))
    })

    test(('should set task color input value'), () => {
        expect(wrapper.find('.form__item--color input').prop('value')).toBe(item.color)
    })

    test(('should set task description input value'), () => {
        expect(wrapper.find('.form__item--description').prop('value')).toBe(item.description)
    })

    test(('should match snapshot'), () => {
        expect(wrapper).toMatchSnapshot()
    })    
})

describe('Should handle Task Form events:', () => {
    beforeEach(() => {
        wrapper = shallow(
            <Form 
                onParentFormSubmit={onParentFormSubmit}
            />
        )
    })

    test('should set name on name change', () => {
        const value = 'Any name'
        wrapper.find('.form__item--name').simulate('change', {
            target: { value }
        })
        expect(wrapper.find('.form__item--name').prop('value')).toBe(value)
    })
    
    test('should set status on status change', () => {
        const value = 'done'
        wrapper.find('.form__item--status').simulate('change', {
            target: { value }
        })
        expect(wrapper.find('.form__item--status').prop('value')).toBe(value)
    })
    
    test('should set start / end dates on DateRangePicker change', () => {
        const startDate = moment('1994-09-08')
        const endDate = moment('1994-09-20')
        wrapper.find(DateRangePicker).prop('onDatesChange')({ startDate, endDate })
        expect(wrapper.find(DateRangePicker).prop('startDate')).toEqual(startDate)
        expect(wrapper.find(DateRangePicker).prop('endDate')).toEqual(endDate)
    })
    
    test('should set calendar focus on focus change', () => {
        const calendarFocused = 'startDate'
        wrapper.find(DateRangePicker).prop('onFocusChange')(calendarFocused)
        expect(wrapper.find(DateRangePicker).prop('focusedInput')).toBe(calendarFocused)
    })
    
    test('should set description on description change', () => {
        const value = 'Any description'
        wrapper.find('.form__item--description').simulate('change', {
            target: { value }
        })
        expect(wrapper.find('.form__item--description').prop('value')).toBe(value)
    })
    
    test('should render error on invalid form submission', (done) => {
        wrapper.find('form').simulate('submit', { preventDefault: jest.fn() })
        expect(wrapper.find('.error-message').text().length).toBeGreaterThan(0) 
        done()
    })

    test('should submit form on valid form submission', (done) => {
        wrapper = mount(
            <MemoryRouter>
                <Form 
                    item={item}
                    onParentFormSubmit={onParentFormSubmit}
                />
            </MemoryRouter>
        ) 
        wrapper.find('form').simulate('submit', { preventDefault: jest.fn() })
        expect(onParentFormSubmit).toHaveBeenCalledWith(item)
        done()
    })
})