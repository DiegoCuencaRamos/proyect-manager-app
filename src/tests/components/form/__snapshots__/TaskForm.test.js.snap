// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Task Form with empty values 1`] = `
<div
  className="container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form__wrapper"
    >
      <input
        className="form__item--name"
        onChange={[Function]}
        placeholder="Name"
        type="text"
        value=""
      />
      <select
        className="form__item--status"
        onChange={[Function]}
        style={
          Object {
            "flexBasis": "100%",
            "marginRight": "0",
          }
        }
        value=""
      >
        <option
          defaultValue={true}
          value=""
        >
          Select status
        </option>
        <option
          value="todo"
        >
          To do
        </option>
        <option
          value="doing"
        >
          Doing
        </option>
        <option
          value="done"
        >
          Done
        </option>
      </select>
      <withStyles(DateRangePicker)
        anchorDirection="left"
        appendToBody={false}
        block={false}
        calendarInfoPosition="bottom"
        customArrowIcon={null}
        customCloseIcon={null}
        customInputIcon={null}
        dayPickerNavigationInlineStyles={null}
        daySize={39}
        disableScroll={false}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        endDate={"1970-01-01T00:00:00.000Z"}
        endDateId="0"
        endDatePlaceholderText="End Date"
        firstDayOfWeek={null}
        focusedInput={null}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepFocusOnInput={false}
        keepOpenOnDateSelect={false}
        minimumNights={1}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPosition="navPositionTop"
        navPrev={null}
        noBorder={false}
        numberOfMonths={1}
        onClose={[Function]}
        onDatesChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableEndDate": [Function],
            "chooseAvailableStartDate": [Function],
            "clearDates": "Clear Dates",
            "closeDatePicker": "Close",
            "dateIsSelected": [Function],
            "dateIsSelectedAsEndDate": [Function],
            "dateIsSelectedAsStartDate": [Function],
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month.",
            "jumpToPrevMonth": "Move backward to switch to the previous month.",
            "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
            "moveFocusByOneMonth": "Switch months.",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
            "openThisPanel": "Open this panel.",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field.",
            "roleDescription": "datepicker",
            "selectFocusedDate": "Select the date in focus.",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        readOnly={false}
        regular={false}
        renderCalendarInfo={null}
        renderDayContents={null}
        renderMonthElement={null}
        renderMonthText={null}
        renderNavNextButton={null}
        renderNavPrevButton={null}
        renderWeekHeaderElement={null}
        reopenPickerOnClearDates={false}
        required={false}
        screenReaderInputMessage=""
        showClearDates={true}
        showDefaultInputIcon={false}
        small={false}
        startDate={"1970-01-01T00:00:00.000Z"}
        startDateId="0"
        startDatePlaceholderText="Start Date"
        verticalHeight={null}
        verticalSpacing={22}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
      <label
        className="form__item--color"
      >
        Task
         color:
        <input
          onChange={[Function]}
          type="color"
          value=""
        />
      </label>
      <textarea
        className="form__item--description"
        onChange={[Function]}
        placeholder="Description"
        value=""
      />
      <input
        className="form__item--submit button"
        type="submit"
        value="Save task"
      />
    </div>
  </form>
</div>
`;

exports[`should render Task Form with task data: should match snapshot 1`] = `
<div
  className="container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form__wrapper"
    >
      <input
        className="form__item--name"
        onChange={[Function]}
        placeholder="Name"
        type="text"
        value="Task 1"
      />
      <select
        className="form__item--status"
        onChange={[Function]}
        style={
          Object {
            "flexBasis": "100%",
            "marginRight": "0",
          }
        }
        value="todo"
      >
        <option
          defaultValue={true}
          value=""
        >
          Select status
        </option>
        <option
          value="todo"
        >
          To do
        </option>
        <option
          value="doing"
        >
          Doing
        </option>
        <option
          value="done"
        >
          Done
        </option>
      </select>
      <withStyles(DateRangePicker)
        anchorDirection="left"
        appendToBody={false}
        block={false}
        calendarInfoPosition="bottom"
        customArrowIcon={null}
        customCloseIcon={null}
        customInputIcon={null}
        dayPickerNavigationInlineStyles={null}
        daySize={39}
        disableScroll={false}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        endDate={"2021-03-18T23:00:00.000Z"}
        endDateId="1616108400000"
        endDatePlaceholderText="End Date"
        firstDayOfWeek={null}
        focusedInput={null}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepFocusOnInput={false}
        keepOpenOnDateSelect={false}
        minimumNights={1}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPosition="navPositionTop"
        navPrev={null}
        noBorder={false}
        numberOfMonths={1}
        onClose={[Function]}
        onDatesChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableEndDate": [Function],
            "chooseAvailableStartDate": [Function],
            "clearDates": "Clear Dates",
            "closeDatePicker": "Close",
            "dateIsSelected": [Function],
            "dateIsSelectedAsEndDate": [Function],
            "dateIsSelectedAsStartDate": [Function],
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month.",
            "jumpToPrevMonth": "Move backward to switch to the previous month.",
            "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
            "moveFocusByOneMonth": "Switch months.",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
            "openThisPanel": "Open this panel.",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field.",
            "roleDescription": "datepicker",
            "selectFocusedDate": "Select the date in focus.",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        readOnly={false}
        regular={false}
        renderCalendarInfo={null}
        renderDayContents={null}
        renderMonthElement={null}
        renderMonthText={null}
        renderNavNextButton={null}
        renderNavPrevButton={null}
        renderWeekHeaderElement={null}
        reopenPickerOnClearDates={false}
        required={false}
        screenReaderInputMessage=""
        showClearDates={true}
        showDefaultInputIcon={false}
        small={false}
        startDate={"2021-03-14T23:00:00.000Z"}
        startDateId="1615762800000"
        startDatePlaceholderText="Start Date"
        verticalHeight={null}
        verticalSpacing={22}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
      <label
        className="form__item--color"
      >
        Task
         color:
        <input
          onChange={[Function]}
          type="color"
          value="green"
        />
      </label>
      <textarea
        className="form__item--description"
        onChange={[Function]}
        placeholder="Description"
        value="This is my task fixure 1"
      />
      <input
        className="form__item--submit button"
        type="submit"
        value="Save task"
      />
    </div>
  </form>
</div>
`;
